File: frontend/src/api/notifications.ts
import client from './client';
import type { Notification, PaginatedResponse } from '@/types';

export const getNotifications = async (page = 1): Promise<PaginatedResponse<Notification> & { unread_count: number }> => {
  const response = await client.get('/notifications', { params: { page } });
  return response.data;
};

export const getUnreadCount = async (): Promise<{ count: number }> => {
  const response = await client.get('/notifications/unread-count');
  return response.data;
};

export const markAsRead = async (id: string): Promise<void> => {
  await client.put(`/notifications/${id}/read`);
};

export const markAllAsRead = async (): Promise<void> => {
  await client.put('/notifications/read-all');
};

export const deleteNotification = async (id: string): Promise<void> => {
  await client.delete(`/notifications/${id}`);
};

export const connectNotificationsWebSocket = (): WebSocket => {
  const protocol = window.location.protocol === 'https:' ? 'wss:' : 'ws:';
  const token = localStorage.getItem('access_token');
  return new WebSocket(`${protocol}//${window.location.host}/api/v1/notifications/ws?token=${token}`);
};

File: backend/app/api/v1/notifications.py
import json
import uuid

from fastapi import APIRouter, Depends, HTTPException, WebSocket, WebSocketDisconnect, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.models.user import User
from app.schemas.notification import NotificationResponse
from app.services.notifications import (
    get_user_notifications,
    get_unread_count,
    mark_as_read,
    mark_all_as_read,
    delete_notification,
)
from app.utils.deps import get_current_user

router = APIRouter(prefix="/notifications", tags=["notifications"])


@router.get("")
async def list_notifications(
    page: int = 1,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await get_user_notifications(db, current_user.id, page=page)


@router.get("/unread-count")
async def unread_count(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    count = await get_unread_count(db, current_user.id)
    return {"count": count}


@router.put("/{notification_id}/read")
async def mark_notification_read(
    notification_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    success = await mark_as_read(db, notification_id, current_user.id)
    if not success:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Notification not found")
    return {"message": "Marked as read"}


@router.put("/read-all")
async def mark_all_read(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    await mark_all_as_read(db, current_user.id)
    return {"message": "All notifications marked as read"}


@router.delete("/{notification_id}", status_code=status.HTTP_204_NO_CONTENT)
async def remove_notification(
    notification_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    success = await delete_notification(db, notification_id, current_user.id)
    if not success:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Notification not found")


@router.websocket("/ws")
async def notifications_websocket(websocket: WebSocket):
    await websocket.accept()

    token = websocket.query_params.get("token")
    if not token:
        await websocket.close(code=4001, reason="Missing token")
        return

    from app.core.security import verify_token
    user_id = verify_token(token, token_type="access")
    if not user_id:
        await websocket.close(code=4001, reason="Invalid token")
        return

    r = await get_redis()
    pubsub = r.pubsub()
    await pubsub.subscribe(f"notifications:{user_id}")

    try:
        while True:
            try:
                await websocket.receive_text()
            except Exception:
                pass

            pub_message = await pubsub.get_message(ignore_subscribe_messages=True, timeout=0.1)
            if pub_message and pub_message["type"] == "message":
                await websocket.send_text(pub_message["data"])
    except WebSocketDisconnect:
        pass
    finally:
        await pubsub.unsubscribe(f"notifications:{user_id}")
        await r.aclose()
